server/test/write.go

```diff
@@ -733,6 +752,335 @@ var writeCommandTests = []writeCommandTest{
+				User:     "impossible:1",
+			},
+		},
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type:      "user",
+				Relations: map[string]*openfgapb.Userset{},
+			},
+			{
+				Type: "org",
+				Relations: map[string]*openfgapb.Userset{
+					"owner": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"owner": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type: "user",
+								},
+							},
+						},
+					},
+				},
+			}},
+		request: &openfgapb.WriteRequest{
+			Deletes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("org:openfga", "owner", "impossible:1"),
+			},
+			},
+		},
+	},
+	{
+		_name: "Write fails if user field contains a type that does not exist",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type: "org",
+				Relations: map[string]*openfgapb.Userset{
+					"owner": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"owner": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type: "user",
+								},
+							},
+						},
+					},
+				},
+			}},
+		request: &openfgapb.WriteRequest{
+			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("org:openfga", "owner", "impossible:1"),
+			},
+			},
+		},
+		err: serverErrors.InvalidWriteInput,
+	},
+	{
+		_name: "Write fails if user field contains a type that is not allowed by the authorization model (which only allows group:...)",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type:      "user",
+				Relations: map[string]*openfgapb.Userset{},
+			},
+			{
+				Type: "group",
+				Relations: map[string]*openfgapb.Userset{
+					"member": {Userset: &openfgapb.Userset_This{}},
+				},
+			},
+			{
+				Type: "document",
+				Relations: map[string]*openfgapb.Userset{
+					"reader": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"reader": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type: "group",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		request: &openfgapb.WriteRequest{
+			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("document:budget", "reader", "user:abc"),
+			}},
+		},
+		err: serverErrors.InvalidTuple("Object of type user is not allowed to have relation reader with document:budget",
+			tuple.NewTupleKey("document:budget", "reader", "user:abc"),
+		),
+	},
+	{
+		_name: "Write succeeds if user field contains a type that is allowed by the authorization model (which only allows user:...)",
```

Comment:
Can we also add a test case to make sure that Writes have to validate against any one of the directly related user types? All of the tests we have here only define 1 DirectlyRelatedUserTypes, and so it would be nice to add a test with 2 writes that succeeds because both of the writes validate against one of the two DirectlyRelatedUserTypes.